File: 3532. Path Existence Queries in a Graph I.js
/**
 * @param {number} n
 * @param {number[]} nums
 * @param {number} maxDiff
 * @param {number[][]} queries
 * @return {boolean[]}
 */
var pathExistenceQueries = function (
    n,
    nums,
    maxDiff,
    queries
) {

    // Component ID for each index
    const componentId =
        new Array(n).fill(0);

    // Assign component IDs
    for (let index = 1; index < n; index++) {

        if (
            nums[index] -
            nums[index - 1] <=
            maxDiff
        ) {

            componentId[index] =
                componentId[index - 1];

        } else {

            componentId[index] =
                componentId[index - 1] + 1;
        }
    }

    const result = [];

    // Answer each query
    for (const [leftIndex, rightIndex] of queries) {

        result.push(
            componentId[leftIndex] ===
            componentId[rightIndex]
        );
    }

    return result;
};

File: 3532. Path Existence Queries in a Graph I.py
class Solution:
    def pathExistenceQueries(
        self,
        n: int,
        nums: List[int],
        maxDiff: int,
        queries: List[List[int]]
    ) -> List[bool]:

        # Component ID for each index
        component_id = [0] * n

        # Assign component IDs
        for index in range(1, n):

            if nums[index] - nums[index - 1] <= maxDiff:

                component_id[index] = (
                    component_id[index - 1]
                )

            else:

                component_id[index] = (
                    component_id[index - 1] + 1
                )

        result = []

        # Answer each query
        for left_index, right_index in queries:

            result.append(
                component_id[left_index] ==
                component_id[right_index]
            )

        return result
